# Pandas Ep. 1

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/ardhiraka/PFDS_sources/master/nbaallelo.csv')

In [3]:
len(df)

126314

In [4]:
df.shape

(126314, 23)

In [5]:
df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  ...  win_equiv  opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies  ...  40.294830     NYK    Knicks       68   
1            0     NYK   Knicks  ...  41.705170     TRH   Huskies       66   
2            0     CHS    Stags  ...  42.012257     NYK    Knicks       47   
3            0     NYK   Knicks  ...  40.692783     CHS     Stags       63   
4            0     DTF  Falcons  ...  38.864048     WSC  Capitols       50   

   opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0  1300.0000  1306.7233              H            L  0.640065   NaN  
1  1300.0000  1293.2767              A            W  0.359935   NaN  
2  1306.7233  1297.0712              H            W  0.631101   NaN  
3  1300.0000  1309.6521              A            L  0.368899   NaN  
4  1300.0000  1320.3811              H            L  0.640065   NaN  

[5 rows x 23 columns]

In [6]:
pd.set_option("display.max.columns", None)
pd.set_option("display.precision", 2)

In [7]:
df.tail()

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts    elo_i    elo_n  \
126309         100            1     CLE  Cavaliers   82  1723.41  1704.39   
126310         102            1     GSW   Warriors  104  1809.98  1813.63   
126311         101            1     CLE  Cavaliers   91  1704.39  1700.74   
126312         102            1     CLE  Cavaliers   97  1700.74  1692.09   
126313         103            1     GSW   Warriors  105  1813.63  1822.29   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
126309      60.31    GSW   Warriors      103    1790.96    1809.98   
126310      68.01    CLE  Cavaliers       91    1704.39    1700.74   
126311      60.01    GSW   Warriors      104    1809.98    1813.63   
126312      59.29    GSW   Warriors      105    1813.63    1822.29   
126313      68.52    CLE  Cavaliers       97    1700.74    1692.09   

       game_location game_result  forecast notes  
126309             H           L      0.55   NaN  
126310             H           W      0.77   NaN  
126311             A           L      0.23   NaN  
126312             H           L      0.48   NaN  
126313             A           W      0.52   NaN

In [8]:
df.tail(10)

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126304      63153  201506070GSW   NBA        1     2015   6/7/2015   
126305      63153  201506070GSW   NBA        0     2015   6/7/2015   
126306      63154  201506090CLE   NBA        1     2015   6/9/2015   
126307      63154  201506090CLE   NBA        0     2015   6/9/2015   
126308      63155  201506110CLE   NBA        1     2015  6/11/2015   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts    elo_i    elo_n  \
126304          98            1     CLE  Cavaliers   95  1708.17  1716.87   
126305          99            1     GSW   Warriors   93  1806.20  1797.50   
126306         100            1     GSW   Warriors   91  1797.50  1790.96   
126307          99            1     CLE  Cavaliers   96  1716.87  1723.41   
126308         101            1     GSW   Warriors  103  1790.96  1809.98   
126309         100            1     CLE  Cavaliers   82  1723.41  1704.39   
126310         102            1     GSW   Warriors  104  1809.98  1813.63   
126311         101            1     CLE  Cavaliers   91  1704.39  1700.74   
126312         102            1     CLE  Cavaliers   97  1700.74  1692.09   
126313         103            1     GSW   Warriors  105  1813.63  1822.29   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
126304      61.31    GSW   Warriors       93    1806.20    1797.50   
126305      67.03    CLE  Cavaliers       95    1708.17    1716.87   
126306      66.62    CLE  Cavaliers       96    1716.87    1723.41   
126307      61.83    GSW   Warriors       91    1797.50    1790.96   
126308      67.79    CLE  Cavaliers       82    1723.41    1704.39   
126309      60.31    GSW   Warriors      103    1790.96    1809.98   
126310      68.01    CLE  Cavaliers       91    1704.39    1700.74   
126311      60.01    GSW   Warriors      104    1809.98    1813.63   
126312      59.29    GSW   Warriors      105    1813.63    1822.29   
126313      68.52    CLE  Cavaliers       97    1700.74    1692.09   

       game_location game_result  forecast notes  
126304             A           W      0.24   NaN  
126305             H           L      0.76   NaN  
126306             A           L      0.47   NaN  
126307             H           W      0.53   NaN  
126308             A           W      0.45   NaN  
126309             H           L      0.55   NaN  
126310             H           W      0.77   NaN  
126311             A           L      0.23   NaN  
126312             H           L      0.48   NaN  
126313             A           W      0.52   NaN

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   gameorder      126314 non-null  int64  
 1   game_id        126314 non-null  object 
 2   lg_id          126314 non-null  object 
 3   _iscopy        126314 non-null  int64  
 4   year_id        126314 non-null  int64  
 5   date_game      126314 non-null  object 
 6   seasongame     126314 non-null  int64  
 7   is_playoffs    126314 non-null  int64  
 8   team_id        126314 non-null  object 
 9   fran_id        126314 non-null  object 
 10  pts            126314 non-null  int64  
 11  elo_i          126314 non-null  float64
 12  elo_n          126314 non-null  float64
 13  win_equiv      126314 non-null  float64
 14  opp_id         126314 non-null  object 
 15  opp_fran       126314 non-null  object 
 16  opp_pts        126314 non-null  int64  
 17  opp_elo_i      126314 non-nul

In [10]:
df.describe()

gameorder   _iscopy    year_id  seasongame  is_playoffs        pts  \
count  126314.00  126314.0  126314.00   126314.00    126314.00  126314.00   
mean    31579.00       0.5    1988.20       43.53         0.06     102.73   
std     18231.93       0.5      17.58       25.38         0.24      14.81   
min         1.00       0.0    1947.00        1.00         0.00       0.00   
25%     15790.00       0.0    1975.00       22.00         0.00      93.00   
50%     31579.00       0.5    1990.00       43.00         0.00     103.00   
75%     47368.00       1.0    2003.00       65.00         0.00     112.00   
max     63157.00       1.0    2015.00      108.00         1.00     186.00   

           elo_i      elo_n  win_equiv    opp_pts  opp_elo_i  opp_elo_n  \
count  126314.00  126314.00  126314.00  126314.00  126314.00  126314.00   
mean     1495.24    1495.24      41.71     102.73    1495.24    1495.24   
std       112.14     112.46      10.63      14.81     112.14     112.46   
min      1091.64    1085.77      10.15       0.00    1091.64    1085.77   
25%      1417.24    1416.99      34.10      93.00    1417.24    1416.99   
50%      1500.95    1500.95      42.11     103.00    1500.95    1500.95   
75%      1576.06    1576.29      49.64     112.00    1576.06    1576.29   
max      1853.10    1853.10      71.11     186.00    1853.10    1853.10   

        forecast  
count  126314.00  
mean        0.50  
std         0.22  
min         0.02  
25%         0.33  
50%         0.50  
75%         0.67  
max         0.98

In [11]:
df["team_id"].value_counts()

BOS    5997
NYK    5769
LAL    5078
DET    4985
PHI    4533
       ... 
INJ      60
PIT      60
DTF      60
TRH      60
SDS      11
Name: team_id, Length: 104, dtype: int64

In [12]:
df["fran_id"].value_counts()

Lakers          6024
Celtics         5997
Knicks          5769
Warriors        5657
Pistons         5650
Sixers          5644
Hawks           5572
Kings           5475
Wizards         4582
Spurs           4309
Bulls           4307
Pacers          4227
Thunder         4178
Rockets         4154
Nuggets         4120
Nets            4106
Suns            4080
Bucks           4034
Trailblazers    3870
Cavaliers       3810
Clippers        3733
Jazz            3555
Mavericks       3013
Heat            2371
Pelicans        2254
Magic           2207
Timberwolves    2131
Grizzlies       1657
Raptors         1634
Hornets          894
Colonels         846
Squires          799
Spirits          777
Stars            756
Sounds           697
Baltimore        467
Floridians       440
Condors          430
Capitols         291
Olympians        282
Sails            274
Stags            260
Bombers          249
Steamrollers     168
Packers           72
Redskins          65
Rebels            63
Waterloo     

#### Lakers memainkan sebanyak 6042 permainan, tetapi LAL hanya memainkan 5078 permainan, tim lakers mana yang memainkan sisanya?

In [21]:
df.loc[df["fran_id"] == "Lakers", "team_id"].value_counts()

LAL    5078
MNL     946
Name: team_id, dtype: int64

In [32]:
df.loc[df["team_id"]== "MNL", "date_game"]

1136     11/4/1948
1152     11/6/1948
1159     11/9/1948
1170    11/11/1948
1183    11/13/1948
           ...    
9701     3/19/1960
9702     3/20/1960
9709     3/22/1960
9711     3/24/1960
9715     3/26/1960
Name: date_game, Length: 946, dtype: object

In [33]:
df.loc[df["team_id"]== "MNL", "date_game"].min()

'1/1/1949'

In [34]:
df.loc[df["team_id"]== "MNL", "date_game"].max()

'4/9/1959'

In [36]:
df.loc[df["team_id"]== "MNL", "date_game"].agg(("min", "max"))

min    1/1/1949
max    4/9/1959
Name: date_game, dtype: object

In [37]:
df.loc[df["team_id"]== "BOS", "pts"].sum()

626484

# Panda's Data Structures

In [46]:
revenues = pd.Series([5555, 7000, 1980])

In [47]:
revenues

0    5555
1    7000
2    1980
dtype: int64

In [48]:
revenues.values

array([5555, 7000, 1980])

In [49]:
revenues.index

RangeIndex(start=0, stop=3, step=1)

In [44]:
city_revenues = pd.Series(
        [4200, 8000, 6500],
        index=["Amsterdam", "Toronto", "Tokyo"]
)

In [45]:
city_revenues

Amsterdam    4200
Toronto      8000
Tokyo        6500
dtype: int64

In [51]:
city_employee_count = pd.Series({"Amsterdam": 5, "Tokyo": 8})
city_employee_count

Amsterdam    5
Tokyo        8
dtype: int64

In [52]:
city_employee_count.keys()

Index(['Amsterdam', 'Tokyo'], dtype='object')

In [54]:
"Tokyo" in city_employee_count

True

In [55]:
"New York" in city_employee_count

False

# Understanding DataFrame Objects

In [56]:
city_data = pd.DataFrame({
        "revenue": city_revenues,
        "employee_count" : city_employee_count
})

In [57]:
city_data

revenue  employee_count
Amsterdam     4200             5.0
Tokyo         6500             8.0
Toronto       8000             NaN

In [58]:
city_data.index

Index(['Amsterdam', 'Tokyo', 'Toronto'], dtype='object')

In [59]:
city_data.values

array([[4.2e+03, 5.0e+00],
       [6.5e+03, 8.0e+00],
       [8.0e+03,     nan]])

In [60]:
city_data.axes

[Index(['Amsterdam', 'Tokyo', 'Toronto'], dtype='object'),
 Index(['revenue', 'employee_count'], dtype='object')]

In [61]:
city_data.axes[0]

Index(['Amsterdam', 'Tokyo', 'Toronto'], dtype='object')

In [62]:
city_data.axes[1]

Index(['revenue', 'employee_count'], dtype='object')

In [64]:
city_data.keys()

Index(['revenue', 'employee_count'], dtype='object')

In [65]:
df.index

RangeIndex(start=0, stop=126314, step=1)

In [66]:
df.axes

[RangeIndex(start=0, stop=126314, step=1),
 Index(['gameorder', 'game_id', 'lg_id', '_iscopy', 'year_id', 'date_game',
        'seasongame', 'is_playoffs', 'team_id', 'fran_id', 'pts', 'elo_i',
        'elo_n', 'win_equiv', 'opp_id', 'opp_fran', 'opp_pts', 'opp_elo_i',
        'opp_elo_n', 'game_location', 'game_result', 'forecast', 'notes'],
       dtype='object')]

# Mengakses elements series

In [67]:
city_revenues

Amsterdam    4200
Toronto      8000
Tokyo        6500
dtype: int64

In [68]:
city_revenues["Toronto"]

8000

In [69]:
city_revenues[1]

8000

In [70]:
city_revenues[-1]

6500

In [71]:
city_revenues[1:]

Toronto    8000
Tokyo      6500
dtype: int64

In [73]:
import pandas as pd

In [74]:
import numpy as np

In [77]:
colors = pd.Series(
            ["red", "purple", "blue", "green", "yellow"],
            index = [1, 2, 3, 5, 8]
)

In [78]:
colors

1       red
2    purple
3      blue
5     green
8    yellow
dtype: object

In [79]:
colors.loc[1]

'red'

In [82]:
colors.iloc[1]

'purple'

In [81]:
colors.loc[8]

'yellow'

# Mengakses Elemen DataFrame

In [83]:
city_data["revenue"]

Amsterdam    4200
Tokyo        6500
Toronto      8000
Name: revenue, dtype: int64

#### mengindeks data city_data dengan kolom berlabel "revenue"

In [84]:
city_data.revenue

Amsterdam    4200
Tokyo        6500
Toronto      8000
Name: revenue, dtype: int64

In [86]:
toys = pd.DataFrame([
        {"name": "ball", "shape": "sphare"},
        {"name": "Rubik's cube", "shape": "cube"}
])

In [87]:
toys

name   shape
0          ball  sphare
1  Rubik's cube    cube

In [89]:
toys["shape"]

0    sphare
1      cube
Name: shape, dtype: object

In [90]:
toys.shape

(2, 2)